### PR TITLE
feat: contact list

### PR DESCRIPTION
**Pull request by FrolYo**

### CODE REVIEW

File: src/dotnet/Chat.UI.Blazor/Components/ContactContextMenu/ContactContextMenu.razor [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+            <CascadingValue Value="@(context.Data as UserContact)">
+                <ContactContextMenuAddToFavorites />
+                <ContactContextMenuPin/>
+            </CascadingValue>
+        </ChildContent>
+    </ContextMenu>
+</RenderIntoStack>
+
+@code{
+        public const string Id = "contextMenu-contactList-chat";
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/ContactContextMenu/ContactContextMenu.razor:16]: Indent is off?

File: src/dotnet/UI.Blazor/Components/Navbar/LeftPanel/WideLeftPanel.razor [hunk 2]
```diff
@@ -34,9 +35,14 @@
...
+    private void OnClick() {
         if (NavbarUI.AddButtonAction.TryGetValue(NavbarUI.ActiveGroupId, out var addAction)) {
             addAction();
         }
     }
+
+    private void OnSearch(string? criteria) {
+        SearchUI.Criteria.Value = criteria ?? "";
+    }
+
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/Navbar/LeftPanel/WideLeftPanel.razor:47]: We normally don't add extra empty lines in such cases.

File: src/dotnet/UI.Blazor/Services/SearchUI.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+namespace ActualChat.UI.Blazor.Services;
+
+public class SearchUI
+{
+    public SearchUI(IStateFactory stateFactory)
+        => Criteria = stateFactory.NewMutable<string>();
+
+    public IMutableState<string> Criteria { get; }
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Services/SearchUI.cs:8]: Props should go before the .ctor.

File: src/dotnet/Users.Contracts/Backend/IRecentsBackend.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+public interface IRecentsBackend
+{
+    [ComputeMethod]
+    Task<ImmutableHashSet<string>> List(
+        string shardKey,
+        RecentScope scope,
+        int limit,
+        CancellationToken cancellationToken);
+    [CommandHandler]
```
Comment (alexyakunin) [inline src/dotnet/Users.Contracts/Backend/IRecentsBackend.cs:11]: Maybe makes sense to add an extra empty line between compute methods and command handlers.

File: src/dotnet/Users.Service.Migration/Migrations/UsersDbContextModelSnapshot.cs [hunk 5]
```diff
@@ -126,6 +126,42 @@ protected override void BuildModel(ModelBuilder modelBuilder)
                     b.ToTable("kvas_entries");
                 });
 
+            modelBuilder.Entity("ActualChat.Users.Db.DbRecent", b =>
```
Comment (alexyakunin) [inline src/dotnet/Users.Service.Migration/Migrations/UsersDbContextModelSnapshot.cs:129]: Maybe `DbRecentEntry` entity + `RecentEntries` table?

File: src/dotnet/Users.Service.Migration/Migrations/UsersDbContextModelSnapshot.cs [hunk 6]
```diff
@@ -126,6 +126,42 @@ protected override void BuildModel(ModelBuilder modelBuilder)
...
+                        .IsRequired()
+                        .HasColumnType("text")
+                        .HasColumnName("scope");
+
+                    b.Property<string>("ShardKey")
+                        .IsRequired()
+                        .HasColumnType("text")
+                        .HasColumnName("shard_key");
+
+                    b.Property<DateTime>("UpdatedAt")
```
Comment (alexyakunin) [inline src/dotnet/Users.Service.Migration/Migrations/UsersDbContextModelSnapshot.cs:150]: I'd probably use `Version` only here, coz technically it's a timestamp.

File: src/dotnet/Users.Service.Migration/Migrations/20220804134443_Recents.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            migrationBuilder.AddColumn<bool>(
+                name: "is_favorite",
+                table: "user_contacts",
+                type: "boolean",
+                nullable: false,
+                defaultValue: false);
+
+            migrationBuilder.CreateTable(
+                name: "recents",
```
Comment (alexyakunin) [inline src/dotnet/Users.Service.Migration/Migrations/20220804134443_Recents.cs:20]: "recent_entries"?

File: src/dotnet/Users.Service/Controllers/UserContactsController.cs [hunk 8]
```diff
@@ -7,16 +7,21 @@ namespace ActualChat.Users.Controllers;
...
 
     [HttpGet, Publish]
-    public Task<ImmutableArray<UserContact>> GetAll(Session session, CancellationToken cancellationToken)
-        => _service.GetAll(session, cancellationToken);
+    public Task<ImmutableArray<UserContact>> List(
+        Session session,
+        CancellationToken cancellationToken)
+        => Service.List(session, cancellationToken);
+
+    public Task<UserContact?> Change(IUserContacts.ChangeCommand command, CancellationToken cancellationToken)
```
Comment (alexyakunin) [inline src/dotnet/Users.Service/Controllers/UserContactsController.cs:25]: `[HttpPost]` is missing.

File: src/dotnet/Users.Service/Db/DbRecent.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
+using System.ComponentModel.DataAnnotations;
+using Stl.Versioning;
+
+namespace ActualChat.Users.Db;
+
+public class DbRecent : IHasId<string>, IHasVersion<long>, IRequirementTarget
```
Comment (alexyakunin) [inline src/dotnet/Users.Service/Db/DbRecent.cs:6]: `DbRecentEntry`?

File: src/dotnet/Users.Service/Db/DbRecent.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    public DbRecent() { }
+    public DbRecent(Recent recent) => UpdateFrom(recent);
+
+    [Key] public string Id { get; set; } = "";
+    public string ShardKey { get; set; } = null!;
+    public string Key { get; set; } = null!;
+    public string Scope { get; set; } = null!;
+    [ConcurrencyCheck] public long Version { get; set; }
+    public DateTime UpdatedAt {
```
Comment (alexyakunin) [inline src/dotnet/Users.Service/Db/DbRecent.cs:18]: Maybe we should agree we never add this field if there is `Version`...

File: src/dotnet/Users.Service/Db/UsersDbContext.cs [hunk 11]
```diff
@@ -21,6 +21,7 @@ public class UsersDbContext : DbContextBase
     public DbSet<DbUserIdentity<string>> UserIdentities { get; protected set; } = null!;
     public DbSet<DbSessionInfo> Sessions { get; protected set; } = null!;
     public DbSet<DbOperation> Operations { get; protected set; } = null!;
+    public DbSet<DbRecent> Recents { get; protected set; } = null!;
```
Comment (alexyakunin) [inline src/dotnet/Users.Service/Db/UsersDbContext.cs:24]: `RecentEntries`?

File: src/dotnet/Users.Service/Module/UsersServiceModule.cs [hunk 12]
```diff
@@ -144,6 +144,8 @@ public override void InjectServices(IServiceCollection services)
         fusion.AddComputeService<IChatReadPositions, ChatReadPositions>();
         fusion.AddComputeService<IServerKvas, ServerKvas>();
         fusion.AddComputeService<IServerKvasBackend, ServerKvasBackend>();
+        fusion.AddComputeService<IRecents, Recents>();
```
Comment (alexyakunin) [inline src/dotnet/Users.Service/Module/UsersServiceModule.cs:147]: `IRecentEntries`, etc.?

Comment (alexyakunin) [reply src/dotnet/Users.Service/Module/UsersServiceModule.cs:147]: The reason I am a bit in favor of renaming this is that `Recent` is kinda super generic & very short.

File: src/dotnet/Users.Service/UserContacts.cs [hunk 13]
```diff
@@ -2,27 +2,51 @@ namespace ActualChat.Users;
...
 public class UserContacts : IUserContacts
 {
-    private readonly IAuth _auth;
-    private readonly IUserContactsBackend _contactsBackend;
+    private IAuth Auth { get; }
+    private IUserContactsBackend ContactsBackend {get; }
+    private ICommander Commander { get; }
 
-    public UserContacts(IAuth auth, IUserContactsBackend contactsBackend)
+    public UserContacts(
```
Comment (alexyakunin) [inline src/dotnet/Users.Service/UserContacts.cs:9]: Is there lots of spaces or something?

**Review by alexyakunin — CHANGES_REQUESTED** [review]
Hi, overall, great job - I left some comments, but nothing critical.

File: src/dotnet/Chat.UI.Blazor/Components/ContactContextMenu/ContactContextMenu.razor (hunk 1 above)
Comment (FrolYo) [reply src/dotnet/Chat.UI.Blazor/Components/ContactContextMenu/ContactContextMenu.razor:16]: вщту

File: src/dotnet/Users.Service/Db/DbRecent.cs (hunk 10 above)
Comment (FrolYo) [reply src/dotnet/Users.Service/Db/DbRecent.cs:18]: this field is to order recent list, it's not informational only

File: src/dotnet/Users.Service/UserContacts.cs (hunk 13 above)
Comment (FrolYo) [reply src/dotnet/Users.Service/UserContacts.cs]: fixed

File: src/dotnet/UI.Blazor/Services/SearchUI.cs (hunk 3 above)
Comment (FrolYo) [reply src/dotnet/UI.Blazor/Services/SearchUI.cs]: fixed

File: src/dotnet/Users.Service/Db/DbRecent.cs (hunk 9 above)
Comment (FrolYo) [reply src/dotnet/Users.Service/Db/DbRecent.cs]: done

File: src/dotnet/Users.Service/Db/UsersDbContext.cs (hunk 11 above)
Comment (FrolYo) [reply src/dotnet/Users.Service/Db/UsersDbContext.cs]: done

File: src/dotnet/Users.Service/Module/UsersServiceModule.cs (hunk 12 above)
Comment (FrolYo) [reply src/dotnet/Users.Service/Module/UsersServiceModule.cs]: done

File: src/dotnet/UI.Blazor/Components/Navbar/LeftPanel/WideLeftPanel.razor (hunk 2 above)
Comment (FrolYo) [reply src/dotnet/UI.Blazor/Components/Navbar/LeftPanel/WideLeftPanel.razor]: fixed

File: src/dotnet/Users.Contracts/Backend/IRecentsBackend.cs (hunk 4 above)
Comment (FrolYo) [reply src/dotnet/Users.Contracts/Backend/IRecentsBackend.cs]: done

File: src/dotnet/Users.Service.Migration/Migrations/UsersDbContextModelSnapshot.cs (hunk 5 above)
Comment (FrolYo) [reply src/dotnet/Users.Service.Migration/Migrations/UsersDbContextModelSnapshot.cs]: done

**Pull request merged**